### PR TITLE
Firebase

### PR DESCRIPTION
**Pull request by Kobzol**

### CODE REVIEW

File: src/btwa_api/scripts.py [hunk 1]
```diff
@@ -10,3 +11,20 @@ def add_handler(username, password):
     session.add(handler)
     session.commit()
     session.close()
+
+
+@click.command("add-handler")
+@click.argument("username")
+@click.argument("password")
```
Comment (bodik) [inline src/btwa_api/scripts.py]: consider ask for password `getpass.getpass` (does not leave traces in history files) or generate password instead (might also implement basic password policy)

Comment (Kobzol) [reply src/btwa_api/scripts.py]: Good point. This is just a script for local debugging, but I added it nevertheless.

File: setup.cfg [hunk 2]
```diff
@@ -44,6 +46,8 @@ install_requires =
     psycopg2-binary >= 2.8
     bcrypt >= 3.1
     pyodbc >= 4.0
```
Comment (petr-k) [inline setup.cfg:48]: is this dependency necessary?

File: src/btwa_api/app/api/endpoints.py [hunk 3]
```diff
@@ -1,44 +1,70 @@
...
+    status: str
+
+
+@router.get("/get-user/{phone}", response_model=User)
+def get_user(phone: str,
              credentials: HTTPBasicCredentials = Depends(security),
+             firebase: Firebase = Depends(get_firebase),
              db: Database = Depends(get_db)):
+    """Find user by his phone number."""
+    check_handler_auth(db, credentials)
```
Comment (petr-k) [inline src/btwa_api/app/api/endpoints.py:31]: Is there a way to do perform the auth check using a middleware, instead of doing it in each individual handler?

File: src/btwa_api/app/api/endpoints.py [hunk 4]
```diff
@@ -1,44 +1,70 @@
...
+    user = get_or_404(firebase.get_user_by_phone(phone))
+    return User(fuid=user["fuid"], status=user["status"])
+
+
+class ProximityRecord(BaseModel):
+    buid: str
+    start: int
```
Comment (petr-k) [inline src/btwa_api/app/api/endpoints.py:38]: Why do `start` and `end` have different type annotations? Seems off (but I might be mistaken)

File: src/btwa_api/app/db/firebase/firebase.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+from firebase_admin import firestore, storage
+from starlette.requests import Request
+
+from ...config import FIREBASE_STORAGE_BUCKET
+
+MAX_IN_QUERY_LENGTH = 10
+ALLOWED_USER_STATUSES = {"unknown", "infected", "cured"}
+
+
+def get_users_batched(collection, ids):
```
Comment (petr-k) [inline src/btwa_api/app/db/firebase/firebase.py:15]: This is a little bit hard to follow. I believe this might rewritten in terms of, e.g. `chunked` from `more_itertools`: https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked

File: src/btwa_api/app/api/endpoints.py [hunk 6]
```diff
@@ -1,44 +1,70 @@
...
 
 
-@router.get("/register/{guid}")
-def register(guid: str,
+class User(BaseModel):
+    fuid: str
+    status: str
+
+
+@router.get("/get-user/{phone}", response_model=User)
```
Comment (bodik) [inline src/btwa_api/app/api/endpoints.py]: consider passing PHONE argument as POST data rather than GET URL argument. phone number might be considered a PI and thus should not be logged to basic webserver logfiles (possibly processed with logging data analytic services)

File: src/btwa_api/app/db/firebase/firebase.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    def get_proximity(self, fuid: str):
+        proximity = get_user_proximity(self.bucket, fuid)
+        buids = list(set(record["buid"] for record in proximity))
+        nearby_users = {user["buid"]: user for user in get_users_batched(self.users, buids)}
+        for record in proximity:
+            record["user"] = nearby_users[record["buid"]]
+        return proximity
+
+    def change_user_status(self, fuid: str, status: str):
+        assert status in ALLOWED_USER_STATUSES
```
Comment (petr-k) [inline src/btwa_api/app/db/firebase/firebase.py:64]: raising a `ValueError` would be more natural than `AssertionError`

File: src/btwa_api/scripts.py (hunk 1 above)
Comment (bodik) [reply src/btwa_api/scripts.py]: resolved by 750541d

File: src/btwa_api/scripts.py [hunk 8]
```diff
@@ -10,3 +13,19 @@ def add_handler(username, password):
     session.add(handler)
     session.commit()
     session.close()
+
+
+@click.command("add-handler")
```
Comment (petr-k) [inline src/btwa_api/scripts.py]: What/who does the "handler" here refer to, exactly? Can't the name/term be more descriptive or accurate?

**Review by petr-k — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 2 above)
Comment (Kobzol) [reply setup.cfg:48]: It is needed for Azure SQL, we're not sure if that's gonna stay.

File: src/btwa_api/app/api/endpoints.py (hunk 4 above)
Comment (Kobzol) [reply src/btwa_api/app/api/endpoints.py:38]: start and end are both `int`s.

File: src/btwa_api/app/api/endpoints.py (hunk 3 above)
Comment (Kobzol) [reply src/btwa_api/app/api/endpoints.py:31]: Yes, it should be a middleware or a decorator, but there's no point in changing this, because it will probably go away. This PR is just an MVP for initial demonstration.

File: src/btwa_api/app/db/firebase/firebase.py (hunk 5 above)
Comment (Kobzol) [reply src/btwa_api/app/db/firebase/firebase.py:15]: I was looking for some window/chunked function in `itertools` but this seemed quite elegant to me, slices can do it without any libraries :) Seems like an overkill to download another package for this to me.

File: src/btwa_api/scripts.py (hunk 8 above)
Comment (Kobzol) [reply src/btwa_api/scripts.py]: Admin/epidemiologist. I know, the name sucks, but as I stated above, it will probably go away. If not I'll rename it.

File: src/btwa_api/app/db/firebase/firebase.py (hunk 5 above)
Comment (petr-k) [reply src/btwa_api/app/db/firebase/firebase.py:15]: Strongly disagree, that's what libraries are for ;) I'll gladly refactor this if you don't have any objections.

Comment (Kobzol) [reply src/btwa_api/app/db/firebase/firebase.py:15]: Indeed libraries are nice, when you have a sensible and integrated package management system, which Python sadly doesn't have :-( Feel free to refactor it, but as I said, there's a possibility that this code will go into the dumpster in the next few days.

Comment (petr-k) [reply src/btwa_api/app/db/firebase/firebase.py:15]: @bodik ok, what about this:

```python
from more_itertools import chunked
...

def get_users_batched(collection, ids):
    return [
        d.to_dict()
        for id_chunk in chunked(ids, MAX_IN_QUERY_LENGTH)
        for d in collection.where("buid", "in", list(id_chunk)).get()
    ]
```

Given that this is mostly a nitpick, it's obviously your call if you feel it's a readability improvement (I do think it is).

But I'll admit I haven't run it nor added a test for it.

Comment (bodik) [reply src/btwa_api/app/db/firebase/firebase.py:15]: Personally I hate double list comprehension, it twist my mind every time I see it. Since it's just a cosmetics I'd leave the original implementation as is for now.

Comment (petr-k) [reply src/btwa_api/app/db/firebase/firebase.py:15]: Not cosmetics imo, but it's up to you. If you find the double list expression hard to follow (understandable, as it's indeed a weird way of `flatmap`ping, albeit a standard, pythonic one), perhaps there's a middle ground where `chunked` is kept. I find the original implementation unreadable.

**Review by jendakol — APPROVED** [review]

**Comment by jendakol** [comment]
I guess the unresolved discussion is not so severe it should block us in any way, so I'm merging this.
